lib/pos_fee_estimator.go

```diff
@@ -0,0 +1,159 @@
+	for _, block := range sortedPastBlocks {
+		if err := posFeeEstimator.addBlock(block); err != nil {
+			return errors.Wrap(err, "PosFeeEstimator.Init: error adding block to pastBlocksMempool")
+		}
+	}
+	return nil
+}
+
+func (posFeeEstimator *PoSFeeEstimator) AddBlock(block *MsgDeSoBlock) error {
+	return posFeeEstimator.addBlock(block)
+}
+
+func (posFeeEstimator *PoSFeeEstimator) addBlock(block *MsgDeSoBlock) error {
+	// Add all transactions from the block to the pastBlocksTransactionRegister.
+	for _, txn := range block.Txns {
+		mtxn, err := NewMempoolTx(txn, block.Header.TstampNanoSecs, block.Header.Height)
+		if err != nil {
+			return errors.Wrap(err, "PoSFeeEstimator.AddBlock: error creating MempoolTx")
+		}
+		if err = posFeeEstimator.pastBlocksTransactionRegister.AddTransaction(mtxn); err != nil {
+			return errors.Wrap(err, "PoSFeeEstimator.AddBlock: error adding txn to pastBlocksTransactionRegister")
+		}
+	}
+	posFeeEstimator.cachedBlocks = append(posFeeEstimator.cachedBlocks, block)
+	// Sort the cached blocks by height & tstamp just to be safe.
+	posFeeEstimator.sortCachedBlocks()
+	if uint64(len(posFeeEstimator.cachedBlocks)) > posFeeEstimator.numBlocks {
+		// Remove the oldest block.
+		if err := posFeeEstimator.RemoveBlock(posFeeEstimator.cachedBlocks[0]); err != nil {
+			return errors.Wrap(err, "PoSFeeEstimator.AddBlock: error removing block from PoSFeeEstimator")
+		}
+	}
+	return nil
+}
+
+func (posFeeEstimator *PoSFeeEstimator) RemoveBlock(block *MsgDeSoBlock) error {
+	// Remove all transaction from the block from the pastBlocksTransactionRegister.
+	for _, txn := range block.Txns {
+		mtxn, err := NewMempoolTx(txn, block.Header.TstampNanoSecs, block.Header.Height)
+		if err != nil {
+			return errors.Wrap(err, "PoSFeeEstimator.RemoveBlock: error creating MempoolTx")
+		}
+		if err = posFeeEstimator.pastBlocksTransactionRegister.RemoveTransaction(mtxn); err != nil {
+			return errors.Wrap(err, "PoSFeeEstimator.RemoveBlock: error removing txn from pastBlocksTransactionRegister")
+		}
+	}
+	blockHash, err := block.Hash()
+	if err != nil {
+		return errors.Wrap(err, "PoSFeeEstimator.RemoveBlock: error computing blockHash")
+	}
+	// Remove the block from the cached blocks.
+	for ii, cachedBlock := range posFeeEstimator.cachedBlocks {
+		cachedBlockHash, err := cachedBlock.Hash()
+		if err != nil {
+			return errors.Wrap(err, "PoSFeeEstimator.RemoveBlock: error computing cachedBlockHash")
+		}
+		if blockHash.IsEqual(cachedBlockHash) {
+			posFeeEstimator.cachedBlocks = append(posFeeEstimator.cachedBlocks[:ii], posFeeEstimator.cachedBlocks[ii+1:]...)
+			break
+		}
+	}
+	return nil
+}
+
+// TODO: Implement me
+func (posFeeEstimator *PoSFeeEstimator) UpdateGlobalParams(globalParams *GlobalParamsEntry) error {
+	return nil
+}
+
+func (posFeeEstimator *PoSFeeEstimator) sortCachedBlocks() {
+	sort.SliceStable(posFeeEstimator.cachedBlocks, func(ii, jj int) bool {
```

Comment:
nit: always prefer to create a new slice rather than mutating one in place. Makes things a bit safer if at some point the original slice is passed around and assumptions on it change
```suggestion
	posFeeEstimator.cachedBlocks = collections.SliceStable(posFeeEstimator.cachedBlocks, func(ii, jj int) bool {
```